subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformationStep.java

```diff
@@ -60,17 +60,21 @@ public TransformationSubject transform(TransformationSubject subjectToTransform,
             LOGGER.info("Transforming {} with {}", subjectToTransform.getDisplayName(), transformer.getDisplayName());
         }
         ImmutableList<File> primaryInputs = subjectToTransform.getFiles();
-        ArtifactTransformDependenciesInternal dependencies = dependenciesResolver.forTransformer(transformer);
-        ImmutableList.Builder<File> builder = ImmutableList.builder();
-        for (File primaryInput : primaryInputs) {
-            Try<ImmutableList<File>> result = transformerInvoker.invoke(transformer, primaryInput, dependencies, subjectToTransform);
-
-            if (result.getFailure().isPresent()) {
-                return subjectToTransform.transformationFailed(result.getFailure().get());
+        Try<ArtifactTransformDependenciesInternal> dependencies = dependenciesResolver.forTransformer(transformer);
+        return dependencies.getFailure().map(failure -> {
+            return subjectToTransform.transformationFailed(failure);
+        }).orElseGet(() -> {
+            ImmutableList.Builder<File> builder = ImmutableList.builder();
+            for (File primaryInput : primaryInputs) {
+                Try<ImmutableList<File>> result = transformerInvoker.invoke(transformer, primaryInput, dependencies.get(), subjectToTransform);
+
+                if (result.getFailure().isPresent()) {
+                    return subjectToTransform.transformationFailed(result.getFailure().get());
+                }
+                builder.addAll(result.get());
             }
-            builder.addAll(result.get());
-        }
-        return subjectToTransform.transformationSuccessful(builder.build());
+            return subjectToTransform.transformationSuccessful(builder.build());
+        });
```

Comment:
Handling the error first seems unnatural to me. In any case, I've introduced a better way, similar to `Optional. I think here: https://github.com/gradle/gradle/pull/7999/files#diff-4b825c0d09e19a583b0f57e6cc4c0046

Reply:
Shall I introduce the `orElseMapFailure` on my branch? Or shall we consolidate the methods on `master`?